turbopack/crates/turbopack-core/src/module_graph/mod.rs

```diff
@@ -310,37 +449,267 @@ impl SingleModuleGraph {
+        );
+
+        let mut stack = vec![GraphNodeIndex {
+            graph_idx: self.graphs.len() - 1,
+            node_idx: entry,
+        }];
+        let mut visited = HashSet::new();
+        while let Some(node) = stack.pop() {
+            let graph = &graphs[node.graph_idx].graph;
+            let node_weight = get_node!(graphs, node);
+            if visited.insert(node) {
+                let neighbors = iter_neighbors(graph, node.node_idx);
+
+                for (edge, succ) in neighbors {
+                    let succ = GraphNodeIndex {
+                        graph_idx: node.graph_idx,
+                        node_idx: succ,
+                    };
+                    let succ_weight = get_node!(graphs, succ);
+                    let edge_weight = graph.edge_weight(edge).unwrap();
+                    let action = visitor(Some((node_weight, edge_weight)), succ_weight);
+                    if !visited.contains(&succ) && action == GraphTraversalAction::Continue {
+                        stack.push(succ);
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Traverses all reachable edges in topological order. The preorder visitor can be used to
+    /// forward state down the graph, and to skip subgraphs
+    ///
+    /// Use this to collect modules in evaluation order.
+    ///
+    /// Target nodes can be revisited (once per incoming edge).
+    /// Edges are traversed in normal order, so should correspond to reference order.
+    ///
+    /// * `entry` - The entry module to start the traversal from
+    /// * `state` - The state to be passed to the visitors
+    /// * `visit_preorder` - Called before visiting the children of a node.
+    ///    - Receives: (originating &SingleModuleGraphNode, edge &ChunkingType), target
+    ///      &SingleModuleGraphNode, state &S
+    ///    - Can return [GraphTraversalAction]s to control the traversal
+    /// * `visit_postorder` - Called after visiting the children of a node. Return
+    ///    - Receives: (originating &SingleModuleGraphNode, edge &ChunkingType), target
+    ///      &SingleModuleGraphNode, state &S
+    ///    - Can return [GraphTraversalAction]s to control the traversal
+    pub async fn traverse_edges_from_entry_topological<S>(
+        &self,
+        entry: ResolvedVc<Box<dyn Module>>,
+        state: &mut S,
+        mut visit_preorder: impl FnMut(
+            Option<(&'_ SingleModuleGraphModuleNode, &'_ ChunkingType)>,
+            &'_ SingleModuleGraphModuleNode,
+            &mut S,
+        ) -> GraphTraversalAction,
+        mut visit_postorder: impl FnMut(
+            Option<(&'_ SingleModuleGraphModuleNode, &'_ ChunkingType)>,
+            &'_ SingleModuleGraphModuleNode,
+            &mut S,
+        ),
+    ) -> Result<()> {
+        let graphs = self.get_graphs().await?;
+        let top_graph = self.graphs.last().unwrap().await?;
+        let entry = top_graph.get_module(entry)?;
```

Comment:
`entry` might not be from the top_graph. I think it has to search through all graphs